woocommerce/woocommerce
# Add a setting and a filter to limit the data introduced in the customer download logs table

### Changes proposed in this Pull Request:

Currently, whenever users downloads a digital item from one of their orders an entry is always created in the `wp_wc_download_log` table. There's no way to disable this, nor any mechanism to control the size of the table, which can lead to performance problems.

This pull request:

* Introduces a new setting, `woocommerce_file_download_logs_max_entries`, to limit the number of entries that the table will hold. Whenever a new entry is added to the table via `WC_Customer_Download::track_download`, the value of the setting is retrieved, and the table is trimmed down to the value of the setting if needed (the older entries, as specified by the `timestamp` column, are deleted).

The setting is accessible via _WooCommerce - Settings - Products - Downloadable products_ and it has a default value of zero, meaning that the table has no size limit (compatible with the old behavior); the table is trimmed "on the fly" when the setting is saved:

![image](https://github.com/woocommerce/woocommerce/assets/937723/f9ff26d6-a413-48d7-9711-f2cd74c6fe4f)

![image](https://github.com/woocommerce/woocommerce/assets/937723/088bff79-a631-4295-95b0-b351e0adf214)

* Introduces a new filter, `woocommerce_track_customer_download`, that allows to disable the creation of log entries for individual downloads (and/or to use an alternative mechanism to log them).

Two new methods are added in the `WC_Customer_Download_Log_Data_Store` class to support this: `get_download_logs_count` and `trim_download_logs_table`. Additionally, the `DownloadPermissionsAdjuster` class is renamed to `DownloadsUtil` and is moved to the `Internal\Utilities` namespace (that's where the setting change action is handled for the "on the fly" trimming of the table).

### How to test the changes in this Pull Request:

#### Prerequisites

1. Create a downloadable product and purchase it as a customer.
2. Either download the product a few times, or even better, create some fake download log entries using this script or a similar one (you can use the "execute once" functionality of [the code snippets plugin](https://wordpress.org/plugins/code-snippets/) for that): 

```php
// Adjust count and time as needed
$count = 1000;
$time = 1577836800; //2020-01-01

global $wpdb;
for($i=1;$i<=$count;$i++) { 
	$wpdb->insert($wpdb->prefix . 'wc_download_log', ['permission_id'=>1, 'timestamp'=>gmdate( 'Y-m-d H:i:s', $time-$i), 'user_id'=>1, 'user_ip_address'=>'0.0.0.0']);
}
```

#### Testing the new table size limit

3. Download the product and check that the size of the downloads log table grows accordingly, you can run `wp db query "select count(*) from wp_wc_download_log"` to get the current table size, or open the _WooCommerce - Settings - Products - Downloadable products_ page (the table size will display in the description of the new setting) .
4. Go to the corresponding settings page if you haven't yet, the value of the new setting should show as 0. Change it to any value that is smaller than the current size of the table and save. Verify that the new table size is the same as the value of the setting (it could be smaller if the table contained entries with the same timestamp).
5. Download the file a few more times, and verify that the size of the table doesn't change. Also verify that the entries that get deleted are the older ones, you can do that by running this command and checking that the timestamps of the downloads you performed appear at the top: `wp db query "select timestamp from wp_wc_download_log order by timestamp desc"`.

#### Testing the new filter

6. Add the following snippet somewhere (you can use the mentioned code snippets plugin):

```php
add_filter('woocommerce_track_customer_download', function($track, $download, $user_id, $user_ip_address ) {
    $user_id ??= '(null)';
    $user_ip_address ??= '(null)';
    wc_get_logger()->debug("Download for permission {$download->get_id()}, user $user_id, IP $user_ip_address", ['source'=>'downloads_log']);
    return false;
}, 10, 4);
```

7. Download the file a couple more of times.
8. Verify that these downloads did _not_ tracked in the downloads log table, but they have got entries in the regular WooCommerce log (_WooCommerce - Status - Logs_) , on a file named `downloads_log`; these entries include the download permission id, the user id and the user IP address.

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
